Machine ID Docs Refactor: Database Access (#32743)
gravitational/teleport: docs/pages/machine-id/access-guides/databases.mdx

---
title: Machine ID with Database Access
description: How to use Machine ID to access database servers
---

(!docs/pages/includes/machine-id/v2-config-warning.mdx!)

> Teleport Database Access protects and controls access to databases. Machine ID

Reviewer: ```suggestion
Teleport protects and controls access to databases. Machine ID
```

can be used to grant machines secure, short-lived access to these databases.

In this guide, you will configure `tbot` to produce credentials that can be

> used to access a database configured in Teleport Database Access.

Reviewer: ```suggestion
used to access a database configured in Teleport.
```

- `tbot` must already be installed and configured on the machine that will
  access the database. For more information, see the
  [platform guides](../platform-guides.mdx).

## Step 1/4. Configure RBAC

First, Teleport must be configured to allow the credentials produced by the bot
to access the database server and database.

If you have followed an access guide, you will have created a role and granted
the bot the ability to impersonate it already. This role just needs to have the
additional privileges added to it.

Use `tctl edit role/example-bot` to add the following rule to the role:

```yaml
spec:
  allow:
    db_labels:
      '*': '*'
    db_names: [example-db]
    db_users: [alice]
    rules:
      - resources: [db_server, db]
        verbs: [read, list]
```

Replace:

- `example-db` with the name of the database which the bot will be used to
  access.
- `alice` with the name of the user which the bot will use when connecting to
  the database.

This rule will allow the bot to do two things:

- Access the database `example` on any database server (due to the `'*': '*'`
  label selector) as the user `alice`.
- Discover information about database resources in Teleport.

> The `'*': '*'` grants access to any database server configured in Teleport.

Reviewer: ```suggestion
The `'*': '*'` label selector grants access to any database server configured in Teleport.
```

In production, consider restricting the bot's access using a more specific
label selector; see the

> [Database Access RBAC guide](../../database-access/rbac.mdx) for more info.

Reviewer: ```suggestion
[Database Access RBAC guide](../../database-access/rbac/configuring-access) for a full reference of database-related role options.
```

## Step 2/4. Configure a database `tbot` output

Now, `tbot` needs to be configured with an output that will produce the

> credentials needed for Database Access. To do this, the `database` output

Reviewer: ```suggestion
credentials needed for database access. To do this, the `database` output
```

| CockroachDB   | `cockroach`   | Provides `cockroach/node.key`, `cockroach/node.crt`, and `cockroach/ca.crt`. |
| Generic TLS   | `tls`         | Provides `tls.key`, `tls.crt`, and `tls.cas` for generic clients that require specific file extensions. |

Outputs must be configured with a destination. In this example, the `directory`
destination will be used. This will write artifacts to a specified directory on
disk. Ensure that this directory can be written to by the Linux user that
`tbot` runs as, and that it can be read by the Linux user that will be accessing
applications.

Modify your `tbot` configuration to add a `database` output:

```yaml
outputs:
- type: database
  destination:
    type: directory
    path: /opt/machine-id
  # Specify the details of the database you wish to connect to.
  service: example-server
  database: example
  username: alice

>   # Specify a format to use for the output credentials. For most databases,
>   # this configuration field can be omitted. See the table below for the
>   # databases that require a value to be specified here.

Reviewer: ```suggestion
  # Specify a format to use for the output credentials. For most databases,
  # this configuration field can be omitted.
```
Since the table is now above the example YAML

  # format: mongo
```

If operating `tbot` as a background service, restart it. If running `tbot` in
one-shot mode, it must be executed before you attempt to execute the Ansible
playbook.

> ## Step 3/4. Configure the database access local proxy

Reviewer: ```suggestion
## Step 3/4. Configure the local database access proxy
```

Now that `tbot` has produced the database access credentials, a local proxy
should be set up to forward the database connections from your database client

> to the Teleport Proxy through a TLS connection. This is necessary as the TLS

Reviewer: ```suggestion
to the Teleport Proxy Service through a TLS connection. This is necessary as the TLS
```

> connection allows the Teleport Proxy to identify the protocol and intended

Reviewer: ```suggestion
connection allows the Teleport Proxy Service to identify the protocol and intended
```

recipient.

> In addition, this local proxy needs to be running as long as the client needs to

Reviewer: ```suggestion
The local proxy needs to be running as long as the client needs to
```

make connections to the database or as long as a connection is still open.

> 

Reviewer: I would move the example unit file and the paragraph that introduces it here. Otherwise, the preceding five paragraphs make it seem like the local proxy is an implementation detail left to the reader, rather than something we provide. 

I would also incorporate some of the architectural information about the local proxy into the instructions for using it.

Author: [5a3317b](https://github.com/gravitational/teleport/pull/32743/commits/5a3317b472d76a9e94fb2fc0594a33806b91cf78)

# services.
# Requires=tbot.service

[Service]
Type=simple
# Ensure that the teleport user/group exists and has read access to the
# destination directory.
User=teleport
Group=teleport
Restart=on-failure
# Adjust `12345` to any port of your choosing that is free on the local loopback
# interface. Adjust `example-server` to the name of the database service in
# teleport.
ExecStart=/usr/local/bin/tbot -c /etc/tbot.yaml proxy --proxy=proxy.example.com:3080 --destination-dir=/opt/machine-id db --port=12345 example-server
ExecReload=/bin/kill -HUP $MAINPID
PIDFile=/run/tbot-db-proxy.pid
LimitNOFILE=8192

[Install]
WantedBy=multi-user.target
```

This will start a local proxy on port `12345` that can be used to connect
to the `example-server` database server. Be sure to customize the `tbot`
parameters as necessary for your local setup.

Finally, run the following commands to enable and start the local proxy service:

```code
$ sudo systemctl enable tbot-db-proxy
$ sudo systemctl start tbot-db-proxy
$ sudo systemctl status tbot-db-proxy
```

### Authenticated tunnel

Whilst the default behaviour requires the client to use client certificate

> authentication, it is possible to configure an "authenticated tunnel". This will

Reviewer: ```suggestion
authentication, it is possible to configure an **authenticated tunnel**. This will
```
We use bold for keywords instead of quotes

automatically attach credentials to any incoming connection to the local port.
Whilst this is less secure, it can be necessary if the client you

> wish to use with Database Access does not support client certificate

Reviewer: ```suggestion
wish to use with your database does not support client certificate
```

authentication.

If you wish to enable the authenticated tunnel mode, `--tunnel` flag is used
with `tbot proxy db...`.

> If you are executing this in the foreground, simply

Reviewer: ```suggestion
If you are executing this in the foreground,
```

provide the `--tunnel` flag. If you are using a systemd service, add `--tunnel`

> to the `ExecStart` in `machine-id-proxy.service` and then reload

Reviewer: ```suggestion
to the `ExecStart` in `machine-id-proxy.service` and then reload the unit.
```
